portfolio/src/main/client/src/util/html.ts

```diff
@@ -0,0 +1,153 @@
+import type {Arrunk} from './types';
+
+type InterpolationValue = Arrunk<string | Node | null | undefined> | EventListener;
+
+/**
+ * Helper function to iterate over all descendants of a node.
+ * @param node The node to iterate over.
+ */
+function* descendants(node: Node): Generator<Node> {
+  for (const child of node.childNodes) {
+    yield child;
+    for (const grandchild of descendants(child)) {
+      yield grandchild;
+    }
+  }
+}
+
+/**
+ * @param arr The array to flatten.
+ * @returns The flattened array.
+ */
+function flatten<T>(arr: Arrunk<T>[]): T[] {
+  const newArr = [];
+  for (const item of arr) {
+    if (Array.isArray(item)) {
+      newArr.push(...item);
+    } else {
+      newArr.push(item);
+    }
+  }
+  return newArr;
+}
+
+// eslint-disable-next-line valid-jsdoc
+/**
+ * Helper function to parse HTML into multiple DOM nodes. Meant to be used
+ * as a template literal tag:
+ * ```
+ * const myElem = htmls`<td>this is the html</td><td>this is the html</td>`;
+ * ```
+ *
+ * You can interpolate HTML elements into the template literal, and it will
+ * insert them properly:
+ * ```
+ * const btn = document.createElement('button');
+ * btn.addEventListener(() => ...);
+ * const myElem = htmls`<p>some text<p>${btn}<p>with a button inside</p>`;
+ * ```
+ *
+ * You can also interpolate arrays of HTML elements or strings.
+ *
+ * Finally, you can interpolate functions, but only to use them as event handlers
+ * in @-attributes. Using them anywhere else is undefined behaviour.
+ * ```
+ * const myHandler = () => console.log('hi');
+ * const myElem = htmls`<button @click='${myHandler}'>Say hi</button>`;
+ * ```
+ * @returns The HTML elements.
+ */
+export function htmlFragment(
+  fragments: TemplateStringsArray,
+  ...items: InterpolationValue[]): Node[] {
+  // combine items first so that arrays can be flattened
+  // without messing up indices
+  const combined: InterpolationValue[] = [fragments[0]];
+
+  for (let i = 0; i < items.length; i++) {
+    combined.push(items[i]);
+    combined.push(fragments[i + 1]);
+  }
+
+  const flattened = flatten(combined);
```

Comment:
Since this is flattened, it's no longer an `Arrunk`, but just an `Array<Node|string|undefined|null>`. It would be nice to make that type assert here. I'd suggest making a file level type for `Node|string|undefined|null` for readability so this could become:
```
const flattened: Renderable[] = flatten(combined);
```
```